File: ports/laravel/src/Miss.php
<?php

namespace I18nKeyless\Laravel;

/**
 * One source string that had no translation, with the languages it was
 * requested in. Sent to POST /translate once per (namespace, key, context).
 */
final class Miss
{
    /**
     * @param  list<string>  $langs  i18n-keyless codes the string was requested in
     */
    public function __construct(
        public readonly string $key,
        public readonly ?string $context,
        public readonly string $namespace,
        public array $langs = [],
    ) {
    }

    /** The lookup key, stored exactly like the SDKs: "key__context". */
    public function lookupKey(): string
    {
        return self::lookupKeyFor($this->key, $this->context);
    }

    public static function lookupKeyFor(string $key, ?string $context): string
    {
        return ($context === null || $context === '') ? $key : "{$key}__{$context}";
    }

    /** Dedupe id: one POST per (namespace, key, context), whatever the languages. */
    public function id(): string
    {
        return $this->namespace.':'.$this->lookupKey();
    }

    public function addLang(string $lang): void
    {
        if (! in_array($lang, $this->langs, true)) {
            $this->langs[] = $lang;
        }
    }

    /** @return array{key: string, context: ?string, namespace: string, langs: list<string>} */
    public function toArray(): array
    {
        return [
            'key' => $this->key,
            'context' => $this->context,
            'namespace' => $this->namespace,
            'langs' => array_values($this->langs),
        ];
    }

    /** @param  array{key: string, context?: ?string, namespace?: string, langs?: list<string>}  $data */
    public static function fromArray(array $data): self
    {
        return new self(
            $data['key'],
            $data['context'] ?? null,
            $data['namespace'] ?? KeylessTranslator::DEFAULT_NAMESPACE,
            array_values($data['langs'] ?? []),
        );
    }
}

File: ports/rails/lib/i18n_keyless/miss.rb
# frozen_string_literal: true

module I18nKeyless
  # One source string that had no translation, with the languages it was
  # requested in. Sent to POST /translate once per (namespace, key, context).
  class Miss
    attr_reader :key, :context, :namespace, :langs

    def initialize(key, context = nil, namespace = Translator::DEFAULT_NAMESPACE, langs = [])
      @key = key
      @context = context.to_s.empty? ? nil : context
      @namespace = namespace.to_s.empty? ? Translator::DEFAULT_NAMESPACE : namespace
      @langs = Array(langs).dup
    end

    # The lookup key, stored exactly like the SDKs: "key__context".
    def lookup_key
      self.class.lookup_key_for(key, context)
    end

    def self.lookup_key_for(key, context)
      context.nil? || context.to_s.empty? ? key.to_s : "#{key}__#{context}"
    end

    # Dedupe id: one POST per (namespace, key, context), whatever the languages.
    def id
      "#{namespace}:#{lookup_key}"
    end

    def add_lang(lang)
      @langs << lang unless @langs.include?(lang)
    end

    # A plain Hash (string keys), for an ActiveJob argument.
    def to_h
      { "key" => key, "context" => context, "namespace" => namespace, "langs" => langs.dup }
    end

    def self.from_h(data)
      data = data.transform_keys(&:to_s)
      new(data["key"], data["context"], data["namespace"] || Translator::DEFAULT_NAMESPACE, data["langs"] || [])
    end
  end
end
